### PR TITLE
Lvrubym 300: As User I want to be able to reset my password when I forgot it

### PR DESCRIPTION
**Pull request by JulkaIII**
## JIRA

* [Main JIRA ticket](http://ssu-jira.softserveinc.com/browse/LVRUBYM-300)
* [JIRA subtask](http://ssu-jira.softserveinc.com/browse/LVRUBYM-301)
* [JIRA subtask](http://ssu-jira.softserveinc.com/browse/LVRUBYM-302)
* [JIRA subtask](http://ssu-jira.softserveinc.com/browse/LVRUBYM-303)
* [JIRA subtask](http://ssu-jira.softserveinc.com/browse/LVRUBYM-304)
* [JIRA subtask](http://ssu-jira.softserveinc.com/browse/LVRUBYM-305)
* [JIRA subtask](http://ssu-jira.softserveinc.com/browse/LVRUBYM-348)

## Code reviewers

- [x] @Mykhailo123
- [x] @HelenMaherovska
- [x] @andriyarterchuk

### Second Level Review

- [ ] @omaik

## Summary of issue

Need to reset user's password when he forgot it

## Summary of change

- Added configurations for mailer;
- Added controller endpoint for reset password request;
- Configured background jobs for application using sidekiq;
- Added background job for reset password email send;
- Customed template for reset password mail, with included internationalization;
- Added rspec tests for PasswordsController.

## Testing approach

- Added rspec tests for PasswordsController.

## CHECK LIST
- [ ]  СI passed
- [ ]  Сode coverage >=90%
- [ ]  PR is reviewed manually again (to make sure you have 100% ready code)
- [ ]  All reviewers agreed to merge the PR
- [ ]  I've checked new feature as logged in and logged out user if needed
- [ ]  PR meets all conventions

### CODE REVIEW

File: spec/models/user_spec.rb [hunk 1]
```diff
@@ -75,4 +75,48 @@
...
+      {
+        'password' => '7654321',
+        'password_confirmation' => '7654321'
+      }
+    end
+
+    context 'when reset_password_token is valid' do
+      it 'resets password' do
+        user.send(:set_reset_password_token)
+        user.reset_password_period_valid?
```
Comment (Mykhailo123) [inline spec/models/user_spec.rb]: Maybe, delete this line, I think it is not nesessery.

File: spec/models/user_spec.rb [hunk 2]
```diff
@@ -75,4 +75,48 @@
...
+        expect(user).to receive(:reset_password)
+        user.reset_pass(params)
+      end
+    end
+    context 'when reset_password_token is not valid' do
+      it "doesn't reset password" do
+        user.send(:set_reset_password_token)
+        Timecop.freeze(Date.today + 7) do
+          user.reset_password_period_valid?
```
Comment (Mykhailo123) [inline spec/models/user_spec.rb]: Maybe, delete this line, I think it is not nesessery.

File: app/controllers/passwords_controller.rb [hunk 3]
```diff
@@ -0,0 +1,33 @@
+class PasswordsController < ApplicationController
+  skip_before_action :authenticate_user!
+
+  def forgot
+    user = User.find_by(email: user_params[:email])
+    if user
+      user.send_reset_info
+      render json: { status: 'ok' }, status: :ok
```
Comment (omaik) [inline app/controllers/passwords_controller.rb]: Redundant json

File: app/controllers/passwords_controller.rb [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+    else
+      render json: { error: 'Email not found' }, status: :not_found
```
Comment (omaik) [inline app/controllers/passwords_controller.rb]: I think user not found, not Email

File: app/controllers/passwords_controller.rb [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+    end
+  end
+
+  def reset
+    user = User.find_by(reset_password_token: user_params[:reset_token])
+    if user
+      if user.reset_pass(user_params)
+        render json: { status: 'ok' }, status: :ok
```
Comment (omaik) [inline app/controllers/passwords_controller.rb]: same here

File: app/mailers/reset_password_mailer.rb [hunk 6]
```diff
@@ -0,0 +1,10 @@
+class ResetPasswordMailer < ApplicationMailer
+  default from: 'takecare.user@gmail.com'
+  default template_path: 'v1/users/mailer'
+
+  def reset_password_instructions(user)
+    @user = user
+    mail(to: @user.email,
+         subject: I18n.t('reset_password_instructions_subject'))
```
Comment (omaik) [inline app/mailers/reset_password_mailer.rb]: You are using I18n for mailer but not for the controller :)

File: app/views/v1/users/mailer/reset_password_instructions.html.erb [hunk 7]
```diff
@@ -0,0 +1,8 @@
+<h3><%= t('greeting') %>, <%= @user.first_name %>!</h3>
+
+<p><%= t('instruction') %></p>
+
+<p><%= link_to t('action'), "http://localhost:1111/reset_password?reset_password_token=#{@user.reset_password_token}" %></p>
```
Comment (omaik) [inline app/views/v1/users/mailer/reset_password_instructions.html.erb]: Make it flexible to take base url based on environment

File: config/environments/development.rb [hunk 8]
```diff
@@ -36,4 +36,16 @@
...
   config.file_watcher = ActiveSupport::EventedFileUpdateChecker
+
+  config.action_mailer.default_url_options = { :host => 'localhost:3000' }
+  config.action_mailer.delivery_method = :smtp
+  config.action_mailer.smtp_settings = {
+    address:              'smtp.gmail.com',
+    port:                 587,
+    domain:               'gmail.com',
+    user_name:            'takecare.user@gmail.com',
+    password:             'user12345',
```
Comment (omaik) [inline config/environments/development.rb]: This should be in ENV variable

File: config/environments/production.rb [hunk 9]
```diff
@@ -43,4 +43,15 @@
...
   # Do not dump schema after migrations.
   config.active_record.dump_schema_after_migration = false
+
+  config.action_mailer.delivery_method = :smtp
+  config.action_mailer.smtp_settings = {
+    address:              'smtp.gmail.com',
+    port:                 587,
+    domain:               'gmail.com',
+    user_name:            'takecare.user@gmail.com',
+    password:             'user12345',
```
Comment (omaik) [inline config/environments/production.rb]: same here

File: config/locales/en.json [hunk 10]
```diff
@@ -0,0 +1 @@
+{"greeting":"Hello","reset_password_instructions_subject":"TakeCare Reset password instructions","instruction":"Someone has requested a link to change your password. You can do this through the link below.","action":"Change my password","instruction_2":"If you didn't request this, please ignore this email.","instruction_3":"Your password won't change until you access the link above and create a new one."}
```
Comment (omaik) [inline config/locales/en.json]: Why do you need JSON?

File: config/routes.rb [hunk 11]
```diff
@@ -2,6 +2,9 @@
   post '/signin', to: 'authentication#authenticate'
   get '/refresh_token', to: 'authentication#refresh_token'
 
+  post 'forgot_password', to: 'passwords#forgot'
+  post 'reset_password', to: 'passwords#reset'
```
Comment (omaik) [inline config/routes.rb]: I think better urls would be 'passwords/reset' and 'passwords/forgot'

File: spec/controllers/passwords_controller_spec.rb [hunk 12]
```diff
@@ -0,0 +1,88 @@
+require 'rails_helper'
+
+RSpec.describe PasswordsController, type: :controller do
+  let(:user) { create(:user) }
+  let!(:token) { user.send(:set_reset_password_token) }
```
Comment (omaik) [inline spec/controllers/passwords_controller_spec.rb]: Any way to omit this? This method is invoked inside model, so it should work well

File: spec/models/user_spec.rb (hunk 1 above)
Comment (omaik) [inline spec/models/user_spec.rb]: +1

File: spec/models/user_spec.rb (hunk 2 above)
Comment (omaik) [inline spec/models/user_spec.rb]: +1

File: .rubocop.yml [hunk 13]
```diff
@@ -5,6 +5,7 @@ AllCops:
   Exclude:
     - 'app/views/**/*'
     - 'db/schema.rb'
+    - 'config/environments/*'
```
Comment (JulkaIII) [inline .rubocop.yml]: Because of too many lines in production and development block

File: spec/models/user_spec.rb (hunk 1 above)
Comment (JulkaIII) [inline spec/models/user_spec.rb]: done

File: spec/models/user_spec.rb (hunk 2 above)
Comment (JulkaIII) [inline spec/models/user_spec.rb]: done

File: config/environments/development.rb (hunk 8 above)
Comment (JulkaIII) [inline config/environments/development.rb]: done

File: app/views/v1/users/mailer/reset_password_instructions.html.erb (hunk 7 above)
Comment (JulkaIII) [inline app/views/v1/users/mailer/reset_password_instructions.html.erb]: done

File: config/environments/production.rb (hunk 9 above)
Comment (JulkaIII) [inline config/environments/production.rb]: done

File: app/views/v1/users/mailer/reset_password_instructions.html.erb (hunk 7 above)
Comment (JulkaIII) [inline app/views/v1/users/mailer/reset_password_instructions.html.erb]: done

File: config/routes.rb (hunk 11 above)
Comment (JulkaIII) [inline config/routes.rb]: done

File: spec/controllers/passwords_controller_spec.rb (hunk 12 above)
Comment (JulkaIII) [inline spec/controllers/passwords_controller_spec.rb]: done

File: app/mailers/reset_password_mailer.rb (hunk 6 above)
Comment (JulkaIII) [inline app/mailers/reset_password_mailer.rb]: done

File: .rubocop.yml (hunk 13 above)
Comment (omaik) [inline .rubocop.yml]: Then exclude it only from block length

File: npm-debug.log [hunk 14]
```diff
@@ -0,0 +1,20 @@
+0 info it worked if it ends with ok
```
Comment (omaik) [inline npm-debug.log]: Ouch

File: app/mailers/reset_password_mailer.rb (hunk 6 above)
Comment (omaik) [inline app/mailers/reset_password_mailer.rb]: Contact Dmytro about I18n, you are using different styles.

File: db/migrate/20170331151946_create_users.rb [hunk 15]
```diff
@@ -6,6 +6,8 @@ def change
       t.integer :gender
       t.string :phone
       t.string :type
+      t.string   :reset_password_token
```
Comment (vikabelous) [inline db/migrate/20170331151946_create_users.rb]: @omaik shouldn't we add new migration instead​ of changing the one that was already merged?

File: app/mailers/reset_password_mailer.rb (hunk 6 above)
Comment (JulkaIII) [inline app/mailers/reset_password_mailer.rb]: done

File: .rubocop.yml (hunk 13 above)
Comment (JulkaIII) [inline .rubocop.yml]: done

File: db/migrate/20170331151946_create_users.rb (hunk 15 above)
Comment (omaik) [inline db/migrate/20170331151946_create_users.rb]: Yes

Comment (omaik) [inline db/migrate/20170331151946_create_users.rb]: Actually, we do not have some real data on production, but let imagine that second demo was release :)

File: app/mailers/reset_password_mailer.rb [hunk 16]
```diff
@@ -0,0 +1,10 @@
+class ResetPasswordMailer < ApplicationMailer
+  default from: 'takecare.user@gmail.com'
```
Comment (omaik) [inline app/mailers/reset_password_mailer.rb]: Use ENV here as well

File: db/migrate/20170331151946_create_users.rb (hunk 15 above)
Comment (omaik) [inline db/migrate/20170331151946_create_users.rb]: Still not fixed

File: spec/models/user_spec.rb [hunk 17]
```diff
@@ -75,4 +75,46 @@
...
+        user.send(:set_reset_password_token)
+        expect(user).to receive(:reset_password)
+        user.reset_pass(params)
+      end
+    end
+    context 'when reset_password_token is not valid' do
+      it "doesn't reset password" do
+        user.send(:set_reset_password_token)
+        Timecop.freeze(Date.today + 7) do
+          expect(user).not_to receive(:reset_password)
```
Comment (omaik) [inline spec/models/user_spec.rb]: `to_not` looks better

File: spec/models/user_spec.rb [hunk 18]
```diff
@@ -75,4 +75,46 @@
...
+    it 'makes a job send reset password instruction email' do
+      user = create(:user)
+
+      expect(ResetPasswordWorker).to receive(:perform_async).with(user.id)
+      user.send_reset_info
+    end
+  end
+
+  describe '#reset_pass' do
+    let!(:user) { create(:user) }
```
Comment (omaik) [inline spec/models/user_spec.rb]: this is redundant

File: spec/models/user_spec.rb [hunk 19]
```diff
@@ -75,4 +75,46 @@
       end
     end
   end
+
+  describe '#send_reset_info' do
+    let(:reset_token) { create(:user).send(:set_reset_password_token) }
```
Comment (omaik) [inline spec/models/user_spec.rb]: seems like this let is not used

File: spec/models/user_spec.rb (hunk 17 above)
Comment (JulkaIII) [inline spec/models/user_spec.rb]: done

File: app/mailers/reset_password_mailer.rb (hunk 16 above)
Comment (JulkaIII) [inline app/mailers/reset_password_mailer.rb]: done

File: spec/models/user_spec.rb (hunk 18 above)
Comment (JulkaIII) [inline spec/models/user_spec.rb]: why? In it block I use 'user' :
```
it 'resets password' do
        user.send(:set_reset_password_token)
        expect(user).to receive(:reset_password)
        user.reset_pass(params)
 end
```

File: spec/models/user_spec.rb (hunk 19 above)
Comment (JulkaIII) [inline spec/models/user_spec.rb]: why? With this let I generate reset password token and then in 'it' block I check if it's generated.

Comment (omaik) [inline spec/models/user_spec.rb]: It is not invoked, so code inside it was not executed

File: spec/models/user_spec.rb (hunk 18 above)
Comment (omaik) [inline spec/models/user_spec.rb]: I mean `let!`

File: db/migrate/20170331151946_create_users.rb (hunk 15 above)
Comment (JulkaIII) [inline db/migrate/20170331151946_create_users.rb]: done

File: spec/models/user_spec.rb (hunk 18 above)
Comment (JulkaIII) [inline spec/models/user_spec.rb]: done

File: spec/models/user_spec.rb (hunk 19 above)
Comment (JulkaIII) [inline spec/models/user_spec.rb]: I edited let and the next 'it' block

File: app/controllers/passwords_controller.rb [hunk 20]
```diff
@@ -0,0 +1,39 @@
+class PasswordsController < ApplicationController
+  skip_before_action :authenticate_user!
+  before_action :find_user, only: [:reset]
```
Comment (omaik) [inline app/controllers/passwords_controller.rb]: This before action pretty confusing. In one action(forgot) you searching user, for other action you are using `before_action`

File: app/models/user.rb [hunk 21]
```diff
@@ -15,4 +15,19 @@ class User < ApplicationRecord
...
+  def send_reset_password_instructions
+    ResetPasswordMailer.reset_password_instructions(self).deliver
+  end
+
+  def send_reset_info
+    set_reset_password_token
+    ResetPasswordWorker.perform_async(id)
+  end
+
+  def reset_pass(params)
```
Comment (omaik) [inline app/models/user.rb]: better idea to call this method reset_password, and call `super` inside it

File: app/models/user.rb [hunk 22]
```diff
@@ -15,4 +15,19 @@ class User < ApplicationRecord
             length: { in: 7..10 },
             format: { with: VALID_PHONE_REGEX, message: 'use only digits' },
             allow_nil: true
+
+  def send_reset_password_instructions
```
Comment (omaik) [inline app/models/user.rb]: For what purpose this method?

Comment (omaik) [inline app/models/user.rb]: This method seems redundant, you can call mailer inside background worker

File: spec/controllers/passwords_controller_spec.rb [hunk 23]
```diff
@@ -0,0 +1,90 @@
+require 'rails_helper'
+
+RSpec.describe PasswordsController, type: :controller do
+  let(:user) do
+    user = create(:user)
+    user.send(:set_reset_password_token)
```
Comment (omaik) [inline spec/controllers/passwords_controller_spec.rb]: any reason for this?

File: spec/models/user_spec.rb [hunk 24]
```diff
@@ -75,4 +75,48 @@
       end
     end
   end
+
+  describe '#send_reset_info' do
+    let(:user) do
+      user = create(:user)
+      user.send(:set_reset_password_token)
```
Comment (omaik) [inline spec/models/user_spec.rb]: same here

File: app/models/user.rb (hunk 22 above)
Comment (JulkaIII) [inline app/models/user.rb]: done

File: spec/controllers/passwords_controller_spec.rb (hunk 23 above)
Comment (JulkaIII) [inline spec/controllers/passwords_controller_spec.rb]: Yes, I generate reset_password_token for user to check the forgot and reset methods.

File: spec/controllers/passwords_controller_spec.rb [hunk 25]
```diff
@@ -0,0 +1,91 @@
+require 'rails_helper'
+
+RSpec.describe PasswordsController, type: :controller do
+  let(:user) { create(:user) }
+
+  describe 'POST #forgot' do
```
Comment (omaik) [inline spec/controllers/passwords_controller_spec.rb]: Add check fot user to receive `send_reset_info` message.

File: spec/controllers/passwords_controller_spec.rb [hunk 26]
```diff
@@ -0,0 +1,91 @@
...
+                        }
+                      }
+
+        expect(response).to have_http_status(:not_found)
+      end
+    end
+  end
+
+  describe 'POST #reset' do
+    context "when user's token is correct" do
```
Comment (omaik) [inline spec/controllers/passwords_controller_spec.rb]: Same for this, now you are checking only statuses

File: spec/controllers/passwords_controller_spec.rb (hunk 25 above)
Comment (JulkaIII) [inline spec/controllers/passwords_controller_spec.rb]: done

File: spec/controllers/passwords_controller_spec.rb (hunk 26 above)
Comment (JulkaIII) [inline spec/controllers/passwords_controller_spec.rb]: done

File: spec/controllers/passwords_controller_spec.rb [hunk 27]
```diff
@@ -0,0 +1,112 @@
...
+        post :reset, format: :json,
+                     params: {
+                       user: {
+                         reset_token: user.reset_password_token,
+                         password: '1234567',
+                         password_confirmation: '1234567'
+                       }
+                     }
+      end
+      it 'has status 200 ok' do
```
Comment (omaik) [inline spec/controllers/passwords_controller_spec.rb]: What is the difference between tests on lines 56 and 84?

Comment (JulkaIII) [inline spec/controllers/passwords_controller_spec.rb]: done, removed unnecessary one)

File: config/routes.rb [hunk 28]
```diff
@@ -2,17 +2,25 @@
...
+
     resources :presence_reports
     resources :report_times
+
     resources :bottle_reports, only: [:index] do
       resources :bottles
     end
+
+    # setting the routes tell devise to use your controller
+    devise_for :users, controllers: { passwords: 'passwords' }
```
Comment (omaik) [inline config/routes.rb]: Do you need this? You specified routes explicitly on lines 5-6

File: app/controllers/passwords_controller.rb [hunk 29]
```diff
@@ -0,0 +1,62 @@
...
+    end
+  end
+
+  private
+
+  def user_language
+    I18n.locale = @user.locale
+  end
+
+  def find_user_by_token
```
Comment (omaik) [inline app/controllers/passwords_controller.rb]: The Logic related to locale could be refactored in better way

File: app/controllers/passwords_controller.rb [hunk 30]
```diff
@@ -0,0 +1,54 @@
...
+      check_new_password
+    else
+      render json: { errors: [I18n.t('mailer.errors.invalid_token')] },
+             status: :not_found
+    end
+  end
+
+  private
+
+  def user_language
```
Comment (omaik) [inline app/controllers/passwords_controller.rb]: better name would be `set_user_language`

**Comment by omaik** [comment]
:+1:

**Pull request merged**